File: src/dots/commands/list.py
import typer
from pathlib import Path
from dots.core.config import DotsConfig
from dots.core.resolver import resolve_modules, get_module_variant_info
from dots.ui.output import console

def list_cmd(
    linked: bool = typer.Option(False, "--linked", help="Show linked modules"),
    unlinked: bool = typer.Option(False, "--unlinked", help="Show unlinked modules"),
    broken: bool = typer.Option(False, "--broken", help="Show broken modules"),
    variant: bool = typer.Option(False, "--variant", help="Show variants"),
    bak: bool = typer.Option(False, "--bak", help="Show backup paths"),
):
    """
    List modules or backups with optional filters.
    """
    config = DotsConfig.load()
    all_modules = resolve_modules(config)

    results = set()

    # If no flags are provided, show all module names
    if not any([linked, unlinked, broken, variant, bak]):
        for module_name in all_modules:
            results.add(module_name)

    # Single pass over modules to evaluate all flags
    for module_name, statuses in all_modules.items():
        if linked and any(s.state == "linked" for s in statuses):
            results.add(module_name)

        if unlinked and any(s.state == "pending" for s in statuses):
            results.add(module_name)

        if broken and any(s.state in ("conflict", "unsafe") for s in statuses):
            results.add(module_name)

        if variant:
            vinfo = get_module_variant_info(config, module_name)
            if vinfo and vinfo.has_variants:
                for v in vinfo.variants:
                    results.add(f"{module_name}:{v}")

        if bak:
            for s in statuses:
                bak_path = Path(str(s.destination) + "-backup")
                if bak_path.exists():
                    results.add(str(bak_path))

    for item in sorted(results):
        console.print(item)

File: internal/cli/list.go
package cli

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/Wilberucx/dots/internal/resolver"
	"github.com/Wilberucx/dots/internal/system"
	"github.com/spf13/cobra"
)

func init() {
	listCmd.RunE = func(cmd *cobra.Command, args []string) error {
		return runList(cmd)
	}
}

func runList(cmd *cobra.Command) error {
	cfg, err := loadConfig()
	if err != nil {
		return fmt.Errorf("loading config: %w", err)
	}

	linked := boolFlag(cmd, "linked")
	unlinked := boolFlag(cmd, "unlinked")
	broken := boolFlag(cmd, "broken")
	showVariant := boolFlag(cmd, "variant")
	showBackups := boolFlag(cmd, "backups")

	allModules, err := resolver.ResolveModules(cfg, nil, nil, "")
	if err != nil {
		return fmt.Errorf("resolving modules: %w", err)
	}

	results := make(map[string]bool)

	// If no flags, show all module names
	if !linked && !unlinked && !broken && !showVariant && !showBackups {
		for moduleName := range allModules {
			results[moduleName] = true
		}
	}

	// Single pass over modules
	for moduleName, statuses := range allModules {
		if linked {
			for _, st := range statuses {
				if st.State == resolver.StateLinked {
					results[moduleName] = true
					break
				}
			}
		}

		if unlinked {
			for _, st := range statuses {
				if st.State == resolver.StatePending {
					results[moduleName] = true
					break
				}
			}
		}

		if broken {
			for _, st := range statuses {
				if st.State == resolver.StateConflict || st.State == resolver.StateUnsafe {
					results[moduleName] = true
					break
				}
			}
		}

		if showVariant {
			vInfo, err := resolver.GetModuleVariantInfo(cfg, moduleName)
			if err == nil && vInfo != nil && vInfo.HasVariants {
				for _, v := range vInfo.Variants {
					results[fmt.Sprintf("%s:%s", moduleName, v)] = true
				}
			}
		}
	}

	// Search .orig files outside module loop
	if showBackups {
		home := cfg.HomeDir
		if home == "" {
			home = system.HomeDir()
		}
		filepath.Walk(home, func(path string, info os.FileInfo, err error) error {
			if err != nil {
				return nil
			}
			if strings.HasSuffix(info.Name(), ".orig") {
				results[path] = true
			}
			return nil
		})
	}

	// Sort and print
	var sorted []string
	for item := range results {
		sorted = append(sorted, item)
	}
	sort.Strings(sorted)

	for _, item := range sorted {
		fmt.Println(item)
	}

	return nil
}
